### ISSUE TITLE
Urgent issue with new site just gone live

### ISSUE THREAD

**Issue by madesimplemedia** [concrete5-community-store/community_store#321]
The site is rather slow, but the main issue is that after it has been running for a while is starts throwing errors like this on page with community store list blocks on:
![4x90cach-eerror](https://user-images.githubusercontent.com/10834004/33720836-1b092c74-db5d-11e7-89fb-60bcf9aaa79d.jpg)

**Comment by Mesuva** [comment]
I've not seen an error like this before with concrete5, it's like it's failed to create some sort of caching file.

I would try turning off block caching and seeing if that makes a difference to whether this errors or not.

How many products are you displaying at once?

**Comment by madesimplemedia** [comment]
Thanks again for your help.
I believe I have fixed this.  In the list block I only needed the link to the page, main image, title and price so I removed all the code relating to options/variations and it's running much faster and not throwing errors half the time now.

I think for use in larger projects, it is necessary to do this.

**Comment by Mesuva** [comment]
Can you give me a sense of how many products you were showing at once and how many variations each product would have been trying to handle?

**Comment by madesimplemedia** [comment]
9 per page but the site has maybe 300 products.

**Comment by madesimplemedia** [comment]
Despite trying to streamline the product list and product blocks, I still get the odd cache error in the logs every 30 - 60 mins:
[12-Dec-2017 14:18:12 Europe/London] PHP Warning:  Unterminated comment starting line 16 in /home/www4x90/public_html/application/files/cache/0fea6a13c52b4d47/25368f24b045ca84/38a865804f8fdcb6/57cd99682e939275/18e23c785c6a72cf/ca47fbe20fed45aa.php on line 16

Thing is, if you open that cache file, you see a correctly terminated comment line line 16:
/* Child Type: integer */

So the error appears to be misleading.

**Comment by Mesuva** [comment]
Could this be a PHP bug?  What PHP version are you running?  Are you able to try a lower/higher version?

**Comment by madesimplemedia** [comment]
hmmm well it's on 7 I am very reluctant to turn it down because the Community Store blocks are pretty slow to be honest

**Comment by Mesuva** [comment]
I'm surprised that 7 products are causing such speed problems... how many variations does each product have?

**Comment by madesimplemedia** [comment]
Running it on a more powerful box now and it's nice and fast, still stripped out all the variations code that I didn't need from the product lists.  Also not running Cloudflare now.

**Comment by mnakalay** [comment]
After several tests, I think @madesimplemedia and I have discovered it is most likely not a Community Store issue.

It seems to be a caching issue linked to the fact the site has 2600+ product images.

Every time the error was thrown it was a cached file not being parsed properly because of an unterminated comment or wrong ending like @madesimplemedia noted.

And every time the file didn't have any such things.

And then we saw the same error happening trying to generate the thumbnails from the dashboard by rescanning all the images. The generation would stop midways with an error.

Turns out, every time you use the file function getApprovedVersion(), C5 checks if the file data (file name, path, thumbnail paths...) is cached or not and caches it if necessary.

The way it's cached is the data for all the files is concatenated as one long string and saved in one file. So with 2600 files, mostly images, that was a hell of a long string.

It seems after a while C5 had trouble parsing that string when adding to it or trying to use it.

Since @madesimplemedia says it runs ok on a more powerful system, I suppose there was some weird memory management issue.

**Comment by Mesuva** [comment]
2600+ images is arguably not that many for a site too, especially ones with galleries

**Comment by mnakalay** [comment]
I agree it shouldn't be a problem but in this case, it is not so much the number of files that seem to be the issue. It is the length of the string used to cache all the files references.

It's a 370ko string and although PHP should be able to handle 2Go strings, the Stash parser seems to stop right in the middle of reading the string with a parse error message. And that whole string is read several times. If you show 7 images, the whole string is read 7 times so maybe it's a memory hog I don't know.

There are no weird characters or anything in the string and everything is escaped. There is nothing I can see or think of that would justify failing at parsing the string right in the middle after reading it several times.

For instance when we tried to generate the thumbnails I think it processed 180 files before it failed. So either it processed a file that had a weird filename or something (but it's not the case) or it just gave up maybe?

It's all educated guesses but if it seems to be woking on a system with more resources, it might be a clue.

**Comment by madesimplemedia** [comment]
On the new hosting I had set a cache expiry time of 180 mins.  I noticed on some days when traffic was at peak levels (Thurs/Fri) I would get maybe 4 errors.  This accounted for about 1% of visitors (4 errors to 400 visitors).  Over the weekend when traffic was much lower there were no errors.
So I have reduced the cache expiry time to 120 mins which should hopefully mean with current levels of traffic that the cache doesn't build up that long image string so much that it starts to error.  I'll monitor it.  I didn put this in as a bug on the C5 site too to see what people say.
